agent/containermetadata/manager.go

```diff
@@ -0,0 +1,159 @@
+// at its creation as this information is not present immediately at the agent's startup
+func (manager *metadataManager) SetContainerInstanceARN(containerInstanceARN string) {
+	manager.containerInstanceARN = containerInstanceARN
+}
+
+// Create creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+// Pointer hostConfig is modified directly so there is risk of concurrency errors.
+func (manager *metadataManager) Create(config *docker.Config, hostConfig *docker.HostConfig, taskARN string, containerName string) error {
+	// Create task and container directories if they do not yet exist
+	metadataDirectoryPath, err := getMetadataFilePath(taskARN, containerName, manager.dataDir)
+	// Stop metadata creation if path is malformed for any reason
+	if err != nil {
+		return fmt.Errorf("container metadata create for task %s container %s: %v", taskARN, containerName, err)
+	}
+
+	err = manager.osWrap.MkdirAll(metadataDirectoryPath, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("creating metadata directory for task %s: %v", taskARN, err)
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadataAtContainerCreate(taskARN, containerName)
+	err = manager.marshalAndWrite(metadata, taskARN, containerName)
+	if err != nil {
+		return err
+	}
+
+	// Add the directory of this container's metadata to the container's mount binds
+	// Then add the destination directory as an environment variable in the container $METADATA
+	binds, env := createBindsEnv(hostConfig.Binds, config.Env, manager.dataDirOnHost, metadataDirectoryPath)
+	config.Env = env
+	hostConfig.Binds = binds
+	return nil
+}
+
+// Update updates the metadata file after container starts and dynamic metadata is available
+func (manager *metadataManager) Update(dockerID string, taskARN string, containerName string) error {
+	// Get docker container information through api call
+	dockerContainer, err := manager.client.InspectContainer(dockerID, inspectContainerTimeout)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we do not update a container that is invalid or is not running
+	if dockerContainer == nil || !dockerContainer.State.Running {
+		return fmt.Errorf("container metadata update for task %s container %s: container not running or invalid", taskARN, containerName)
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadata(dockerContainer, taskARN, containerName)
+	err = manager.marshalAndWrite(metadata, taskARN, containerName)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Clean removes the metadata files of all containers associated with a task
+func (manager *metadataManager) Clean(taskARN string) error {
+	metadataPath, err := getTaskMetadataDir(taskARN, manager.dataDir)
+	if err != nil {
+		return fmt.Errorf("clean task %s: %v", taskARN, err)
```

Comment:
please add more context here: `"clean task metadata: unable to get metadata directory for task %s: %v"`

Reply:
fixed.